### ISSUE TITLE
Transformer models above version 4.1.1 can't be wrapped with the configure_interpretable_embedding_layer

### ISSUE THREAD

**Issue by apepa** [pytorch/captum#592]
## 🐛 Bug

When using the latest version of the transformers package, the models cannot be wrapped with the configure_interpretable_embedding_layer function. There is a line in the new package version that gets the shape of the input the following way : `batch_size, seq_length = input_shape`, but the input we pass has three dimensions as we pass the embedding. So we get the '_too many values to unpack (expected 2)_' error. Is it possible to solve this problem with the latest version of the package?

## To Reproduce

You can see the issue on this Colab notebook:

https://colab.research.google.com/drive/1-aZ9-Kzkb_BVb-8vcvHBAAYy2iBk0khV?hl=en#scrollTo=ZC9tY3vsuBjL&uniqifier=2

If you uncomment the `!pip install transformers==4.1.1` line, which installs a lower version, the error is not present.

**Comment by bilalsal** [comment]
Thank you for reporting this issue @apepa!

Captum's `configure_interpretable_embedding_layer` does work [as intended](https://github.com/pytorch/captum/blob/master/captum/attr/_models/base.py#L71) and the problem indeed stems from [the new line](https://github.com/huggingface/transformers/blob/master/src/transformers/models/bert/modeling_bert.py#L920) in you pointed out in `transformers 4.2.x`.
As the line reveals, the new code seemingly unnecessarily assumes the dimensionality of input_ids is exactly 2.

In the earlier version, the code did not make this assumption: [`seq_length = input_shape[1]`]( https://github.com/huggingface/transformers/blob/v4.0.1-release/src/transformers/models/bert/modeling_bert.py#L189)
and  [effective_batch_size = input_shape[0]](https://github.com/huggingface/transformers/blob/v4.0.1-release/src/transformers/models/bert/modeling_bert.py#L1181)

What about you create a PR [in the transformers repo](https://github.com/huggingface/transformers/), suggesting the following changes to L916-L925 of modeling_bert.py in master?

```python
        if input_ids is not None and inputs_embeds is not None:
            raise ValueError("You cannot specify both input_ids and inputs_embeds at the same time")
        elif input_ids is not None:
            input_shape = input_ids.size()
        elif inputs_embeds is not None:
            input_shape = inputs_embeds.size()[:-1]
        else:
            raise ValueError("You have to specify either input_ids or inputs_embeds")
        batch_size = input_shape[0]
        seq_length = input_shape[1]
```

Hope this helps,
Bilal

**Comment by apepa** [comment]
Thank you for looking into this!

I understand that the change they introduced is responsible for the created incompatibility between the two packages.
I'm also not quite sure that the change in the transformers package is conceptually wrong as they have two possible inputs -- input tokens and input embeddings, and providing either matches the correct dimension number, so you might also say that it works as expected. 

I was rather hoping to find a way to provide embeddings, which would be the correct thing to do in this case (this will affect only the transformer-related tutorial). I remember this didn't work for me the first time. I'll give it another try and come back with more information.

**Comment by bilalsal** [comment]
Thank you for exploring alternative avenues, @apepa.

I noticed that if we provide input embeddings instead of input tokens, the latest code in modeling_bert.py does reduce the size by 1 before assigning it to input_shape, unlike the case with input tokens.
  
```python
        elif input_ids is not None:
            input_shape = input_ids.size()
            batch_size, seq_length = input_shape
        elif inputs_embeds is not None:
            input_shape = inputs_embeds.size()[:-1]
            batch_size, seq_length = input_shape
```

Is this always justified?

**Comment by NarineK** [comment]
@apepa, if you want to use IG instead of using LayerIG by setting your custom forward function to something like this:
```
def custom_forward2(inp_emb, attention_mask):
    out = model(inputs_embeds=inp_emb, attention_mask=attention_mask)
    return out.logits
```
and calling IG like this:

attributions = IntegratedeGradients(custom_forward2).attribute(model.bert.embeddings.word_embeddings(encoded_sent_tensor),
                                    baselines=model.bert.embeddings.word_embeddings(ref_sent_tensor),
                                    additional_forward_args=(sent_mask_tensor,),
                                    target=label,
                                    n_steps=5)
```

This is equivalent to:
```
def custom_forward(inputs_id, attention_mask):
    out = model(input_ids=inputs_id, attention_mask=attention_mask)
    return out.logits

attributions, delta = LayerIntegratedGradients(custom_forward, model.bert.embeddings.word_embeddings).attribute(encoded_sent_tensor, additional_forward_args=(sent_mask_tensor,),
                                    baselines = ref_sent_tensor,
                                    target=label,
                                    n_steps=1)
```
The attribution will be w.r.t. the output of thee word embedding layer.

**Comment by bilalsal** [comment]
Thanks @NarineK for the good proposal! The latest packages of transformers can handle the case of `inputs_embeds` without problems.
Should we update the tutorial to adopt the wrapper + inputs_embeds route?

**Comment by NarineK** [comment]
It would be good to update it. There is a case when I look into token, word and position embeddings separately with the configure layer we - might be able to do that with multi-layer IG right now + layer conductance case would need a fix too.

**Comment by apepa** [comment]
Thank you @NarineK for the suggestion. I don't use Integrated Gradients, but showcase other approaches like Saliency and InputXGradient in that tutorial.

I just tried the previous idea using the embeddings parameter and it worked. I did the following:
1) Create a wrapper for the BERT model to pass embeddings instead of the ids (this could also be done with a function wrapper):
```
class BertWrapper(torch.nn.Module):
    def __init__(self, transformer):
        super(BertWrapper, self).__init__()
        self.transformer = transformer
    def forward(self, input, attention_mask=None):
        return self.transformer(inputs_embeds=input, attention_mask=attention_mask)['logits']
```

2) Extract the BERT embeddings for the token ids:
```
input_embeddings = model.transformer.bert.embeddings(input_ids)
```
3) Pass the embeddings as input instead of the tokens:
```
ablator = Saliency(model)
additional_args = input_ != tokenizer.pad_token_id
attributions = ablator.attribute(input_embeddings, additional_forward_args=(additional_args,), target=0)
```

More importantly, this doesn't need the configure_interpretable_embedding_layer/remove_interpretable_embedding_layer functionality and would not even work if you accidentally use it (which was my mistake last time I tried).

I would even recommend using this approach whenever possible -- having the possibility of passing the embeddings or the token_ids to a model instead of the configure_interpretable_embedding_layer/remove_interpretable_embedding_layer functionality, as there's more control on the user side.
